### PR TITLE
Sk passport

### PR DESCRIPTION
**Pull request by jpallen**

### CODE REVIEW

File: app/coffee/Features/Authentication/AuthenticationController.coffee [hunk 1]
```diff
@@ -11,63 +11,106 @@ basicAuth = require('basic-auth-connect')
...
+			return req.session.user
+		else if req?.session?.passport?.user
+			return req.session.passport.user
 		else
-			return callback null, null
+			return null
 
-		UserGetter.getUser query, callback
+	# TODO: perhaps should produce an error if the current user is not present
+	getLoggedInUser: (req, callback = (error, user) ->) ->
```
Comment (henryoswald) [inline app/coffee/Features/Authentication/AuthenticationController.coffee]: This func is only used in GithubSyncController, for not very authentication needs, kill it off instead and just let github sync get the user itself?

File: app/coffee/infrastructure/Server.coffee [hunk 2]
```diff
@@ -87,11 +91,35 @@ webRouter.use csrfProtection
...
+		usernameField: 'email',
+		passwordField: 'password'
+	},
+	AuthenticationController.doPassportLogin
+))
+passport.serializeUser(AuthenticationController.serializeUser)
+passport.deserializeUser(AuthenticationController.deserializeUser)
+
+# standard access to the current user
+currentUserMiddleware = (req, res, next) ->
```
Comment (henryoswald) [inline app/coffee/infrastructure/Server.coffee]: is this used by passport? Can't see us calling it anywhere.

File: app/coffee/Features/Authentication/AuthenticationController.coffee (hunk 1 above)
Comment (ShaneKilkelly) [inline app/coffee/Features/Authentication/AuthenticationController.coffee]: Good call, will do

File: app/coffee/infrastructure/Server.coffee (hunk 2 above)
Comment (ShaneKilkelly) [inline app/coffee/infrastructure/Server.coffee]: Nope, it's dead code, will remove

File: app/coffee/Features/Authentication/AuthenticationController.coffee [hunk 3]
```diff
@@ -126,26 +169,3 @@ module.exports = AuthenticationController =
...
-		lightUser =
-			_id: user._id
-			first_name: user.first_name
-			last_name: user.last_name
-			isAdmin: user.isAdmin
-			email: user.email
-			referal_id: user.referal_id
-			session_created: (new Date()).toISOString()
-			ip_address: req.ip
-		# Regenerate the session to get a new sessionID (cookie value) to
```
Comment (ShaneKilkelly) [inline app/coffee/Features/Authentication/AuthenticationController.coffee]: We do these in `AuthenticationController.serializeUser`

Comment (henryoswald) [inline app/coffee/Features/Authentication/AuthenticationController.coffee]: My selection wasn't great there, I was talking about the "Regenerate the session"

File: app/coffee/Features/StaticPages/UniversityController.coffee [hunk 4]
```diff
@@ -36,7 +37,8 @@ module.exports = UniversityController =
 
 
 	getIndexPage: (req, res)->
-		client = sixpack.client(req?.session?.user?._id?.toString() || req.ip)
+		user = AuthenticationController.getSessionUser(req)
```
Comment (jpallen) [inline app/coffee/Features/StaticPages/UniversityController.coffee]: This could be just user_id directly?

**Comment by jpallen** [comment]
To check manually:

* Changes from web *

- [ ] Something that requires log in
- [ ] Something that uses requireGlobalLogin
- [ ] The restricted page when logged in / not logged in
- [ ] Beta program opt in
- [ ] Beta program op out
- [ ] Send chat message
- [ ] Invite to project
- [ ] Resend invite
- [ ] view invite page
- [ ] Accept invite
- [ ] Compile
- [ ] Stop compile
- [ ] Download aux files or word count
- [ ] Download PDF
- [ ] Get Contacts
- [ ] Invalid CSRF
- [ ] Error page
- [ ] Get notifications
- [ ] Mark notification as read
- [ ] Set project public access level
- [ ] Clone project
- [ ] Create project
- [ ] Project list page 
- [ ] Editor as both logged in and anonymous user
- [ ] Referrals
- [ ] Rate limiting (hard to test properly, but code looks ok. Another pair of eyes to sign of on this one please)
- [ ] Spell check
- [ ] Home page logged in vs not
- [ ] University page
- [ ] Plans page
- [ ] New subscription page
- [ ] User subscription dashboard
- [ ] Custom subscription page
- [ ] Edit billing details page
- [ ] Create subscription
- [ ] Update subscription (change plan)
- [ ] Cancel subscription
- [ ] Reactivate subscription
- [ ] Upgrade to annual plan page
- [ ] Process annual plan upgrade
- [ ] Extend trial (on cancelling I think)
- [ ] Add user to a group
- [ ] Remove user from a group
- [ ] Remove self from a group
- [ ] Render group admin page
- [ ] Start joining a group (send email)
- [ ] Finish joining a group (email verified)
- [ ] Export group CSV
- [ ] Get tags
- [ ] Create, read, update, delete tags
- [ ] Get tracked changes
- [ ] Upload project
- [ ] Upload file
- [ ] Delete user
- [ ] Show settings page
- [ ] Update settings (either in project left menu, e.g. font size, or on project list page, e.g. role)
- [ ] Unsubscribe from newsletter
- [ ] Change password
- [ ] Logout
- [ ] /user/personal_info
- [ ] with cdn blocked
- [ ] locals.getUserEmail() (figure out where this is used a test it)
- [ ] Building of referral URLs and ids
- [ ] Navbar logged in /logged out/ admin

**Comment by jpallen** [comment]
## Web module changes

- [ ] Dropbox on settings page
- [ ] Link dropbox
- [ ] Unlink Dropbox
- [ ] Link Github
- [ ] Unlink Github
- [ ] Import repo from Github
- [ ] Export project to GitHub
- [ ] Send changes from SL to Github
- [ ] Pull changes from Github to SL
- [ ] Register new user
- [ ] Session invalidation on password reset
- [ ] Register form still acts as login

## Templates

- [ ] Create project from template

## References (tpr and search)

- [ ] Link to Mendeley
- [ ] Unlink from Mendeley
- [ ] Import bibtex from Mendeley
- [ ] Search in bibtex

File: app/coffee/Features/Referal/ReferalMiddleware.coffee [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
 module.exports = RefererMiddleware =
 	getUserReferalId: (req, res, next) ->
-		if req.session? and req.session.user?
-			User.findById req.session.user._id, (error, user) ->
-				return next(error) if error?
-				req.session.user.referal_id = user.referal_id
-				next()
+		if AuthenticationController.isUserLoggedIn()
+			user = AuthenticationController.getSessionUser(req)
+			req.user.referal_id = user.referal_id
```
Comment (ShaneKilkelly) [inline app/coffee/Features/Referal/ReferalMiddleware.coffee]: turns out this was dead code, removed.

**Pull request merged**